tests/core/test_shots.py

```diff
@@ -0,0 +1,76 @@
+from datetime import datetime, timezone
+from lagom import Container
+
+from parlant.core.agents import Agent
+from parlant.core.customers import Customer
+from parlant.core.engines.alpha.guideline_proposer import (
+    GuidelineProposer,
+    GuidelinePropositionSchema,
+    GuidelinePropositionShot,
+    GuidelinePropositionsSchema,
+)
+from parlant.core.guidelines import GuidelineContent
+from parlant.core.sessions import Event, EventId
+from parlant.core.shots import ShotCollection
+
+
+async def test_that_appended_shot_is_displayed_in_guideline_proposer_prompt(
+    container: Container,
+    agent: Agent,
+    customer: Customer,
+) -> None:
+    guideline_proposer = container[GuidelineProposer]
+    shot_collection = container[ShotCollection[GuidelinePropositionShot]]
+
+    guideline_content = GuidelineContent(
+        condition="A user compliments the product",
+        action="Tell him to stop being a bitch",
+    )
+
+    guideline_proposition_shot = GuidelinePropositionShot(
+        description="Test Shot Description",
+        interaction_events=[
+            Event(
+                id=EventId("test_id"),
+                source="customer",
+                kind="message",
+                creation_utc=datetime.now(timezone.utc),
+                offset=0,
+                correlation_id="",
+                data={"message": "This is a very cool feature, man!"},
+                deleted=False,
+            )
+        ],
+        guidelines=[guideline_content],
+        expected_result=GuidelinePropositionsSchema(
+            checks=[
+                GuidelinePropositionSchema(
+                    guideline_number=1,
+                    condition=guideline_content.condition,
+                    condition_application_rationale="Rationale: The user said that the feature is cool",
+                    condition_applies=True,
+                    action=guideline_content.action,
+                    guideline_previously_applied="fully",
+                    applies_score=8,
+                )
+            ]
+        ),
+    )
+
+    await shot_collection.append(guideline_proposition_shot)
+
+    shots = await shot_collection.list()
+    prompt = guideline_proposer._format_prompt(
```

Comment:
We shouldn't access underscore-prefixed methods from outside the class: they're considered private.
If you want to access it, you need to make it public—remove the underscore prefix.

HOWEVER -- I wouldn't go through hoops here to format the prompt. It's delving too deep into implementation details, and the assertions you have there are quite fragile anyway.

Instead of using `_format_prompt`, just expose a `guideline_proposer.shots` property which internally simply returns the module's shots (which is the one you're modifying via the container). Then assert that the added shots are contained in it.

Reply:
Since the module's shots is actually the ShotCollection, and we have an async function for list the shots, I made the function of `async def shots(self) -> Sequence[GuidelinePropositionShot]` instead of property (unless you want other option i.e. propetry of shots return the ShotCollection / async_property .. )